### PR TITLE
New interpolation functions to handle terraform commands

### PR DESCRIPTION
**Pull request by jocgir**
This should solve #171.

Add new interpolation functions to list terraform commands that make use of -var, -var-file and -lock-timeout.

`get_terraform_commands_that_need_locking()`
`get_terraform_commands_that_need_vars()`

```hcl
terragrunt = {
  terraform {
    # Force Terraform to keep trying to acquire a lock for up to 20 minutes if someone else already has the lock
    extra_arguments "retry_lock" {
      commands  = "${get_terraform_commands_that_need_locking()}"
      arguments = ["-lock-timeout=20m"]
    }
  }
}
```

Note that the implementation has been a little bit complicated by the fact that we cannot just implement it as suggested:
```hcl
      commands  = ${get_terraform_commands_that_need_locking()}
```

This syntax is not compliant with HCL as interpolation function should be between quotes.
```hcl
      commands  = "${get_terraform_commands_that_need_locking()}"
```

But, since these functions are returning a list of string, we do not want to get this as result:
```hcl
      commands  = "["apply", "plan", ...]"
```

Which is invalid. We expect to get rid of the surrounding quotes and get:
```hcl
      commands  = ["apply", "plan", ...]
```

So, I had to change the return of `resolveTerragruntInterpolation` and `executeTerragruntHelperFunction` to handle `interface{}` instead of `string`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -460,7 +460,7 @@ when creating the S3 bucket or DynamoDB table.
 
 ### Keep your CLI flags DRY
 
-#### Motivation
+#### <a name="extra_arguments"></a>Motivation
```
Comment (brikis98) [inline README.md]: GitHub headers automatically have anchor links. You could just link to #keep-your-cli-flags-dry.

File: README.md [hunk 2]
```diff
@@ -1146,6 +1165,34 @@ terragrunt = {
 
 The common.tfvars located in the terraform root folder will be included by all applications, whatever their relative location to the root.
 
+#### get_terraform_commands_that_need_vars
+
+`get_terraform_commands_that_need_vars()`
+
+Returns the list of terraform commands that accept -var and -var-file parameters. This function is used when defining [extra_arguments](#extra_arguments):
```
Comment (brikis98) [inline README.md]: Replace the `:` with a `.` at the end of the sentence.

Comment (brikis98) [inline README.md]: Or, better yet, add some example code after the `:` :)

File: README.md [hunk 3]
```diff
@@ -497,6 +497,23 @@ Each `extra_arguments` block includes an arbitrary name (in the example above, `
...
+[get_terraform_commands_that_need_locking](#get_terraform_commands_that_need_locking) and
+[get_terraform_commands_that_need_vars](#get_terraform_commands_that_need_vars)
+since they provide the complete list of terraform commands that make use of the desired parameter:
+
+```hcl
+terragrunt = {
+  terraform {
+    # Force Terraform to keep trying to acquire a lock for up to 20 minutes if someone else already has the lock
+    extra_arguments "retry_lock" {
+      commands  = "${get_terraform_commands_that_need_locking()}"
```
Comment (brikis98) [inline README.md]: Would the implementation and docs be simplified if `get_terraform_commands_that_need_locking()` returned a comma-separated list of command names and you the user provided the brackets themselves?

```
commands = ["${get_terraform_commands_that_need_locking()}"]
```

Comment (brikis98) [inline README.md]: That's also more readable.

File: README.md (hunk 1 above)
Comment (jocgir) [inline README.md]: OK, I wasn't sure how to link to title with spaces, thank you.

File: README.md (hunk 2 above)
Comment (jocgir) [inline README.md]: OK for the . I am hesitant to add more example since there is already an example of both commands.

File: README.md (hunk 3 above)
Comment (jocgir) [inline README.md]: That doesn't really simplify the implementation because we still have to generate a comma separated list of quoted strings and still have to get rid of the surrounding quotes around the declaration itself.

This is the technique used by terraform. When a command returns an array, the result is an array and there is no need to add the brackets.The only advantage would be to allow the user to specify more commands, but since the list is complete, there is no real use.

**Comment by jocgir** [comment]
@brikis98, If you don't mind, I found a bug with the current management of extra_arguments. I fixed it, but will include the fix in this PR.

command like `terraform taint -module xxx yyy` which mark the resource yyy of module xxx for refresh is not handling the `extra_arguments` properly.


If terragrunt is configured to inject -lock-timeout on command.
```
terragrunt taint -module xxx yyy
```
will result in:
```
terraform taint -module xxx yyy -lock-timeout=20m
```
which give an error. `terraform` expect to receive it's options between the command `taint` and the arguments. So it expect:
```
terraform taint -module xxx -lock-timeout=20m  yyy 
```
But is is difficult to find the proper insertion point. So the easiest way to do thing is to insert the extra arguments directly after the command:
```
terraform taint -lock-timeout=20m -module xxx yyy 
```
My suggested implementation handle the fact that some `terraform` commands require sub command like `terraform state <sub-command>`.

This also has the advantage of not overwriting a desired behaviour. If we automatically inject `-var region-us-east-1` in the terragrunt configuration.
```
terragrunt apply -var region=us-west-2
```
We get:
```
terraform apply -var region=us-west-2 -var region=us-east-1
```
Which override the explicitly declared value supplied by the user

With the proposed change, we get
```
terraform apply -var region=us-east-1 -var region=us-west-2
```
Which result in a more compliant command to the user intent.

**Comment by brikis98** [comment]
Ah, yes, that's #142. I like your idea for a fix!

**Comment by brikis98** [comment]
@jocgir What is the latest status with this PR?

**Comment by jocgir** [comment]
Hi @brikis98,

If I remind well, I gave made the modifications you proposed and I was waiting for your comment or approval.

File: README.md [hunk 4]
```diff
@@ -1146,6 +1165,59 @@ terragrunt = {
...
+
+_Note: Functions that return a list of values must be used in a single declaration like:_
+
+```hcl
+commands = ["${get_terraform_commands_that_need_vars()}"]
+
+# which result in:
+commands = ["apply", "console", "destroy", "import", "plan", "push", "refresh"]
+
+# They shall not be used to do string composition like:
```
Comment (brikis98) [inline README.md]: s/They shall not/They cannot/`

File: cli/cli_app.go [hunk 5]
```diff
@@ -182,7 +188,17 @@ func runTerragrunt(terragruntOptions *options.TerragruntOptions) error {
...
+		if util.ListContainsElement(TERRAFORM_COMMANDS_WITH_SUBCOMMAND, terragruntOptions.TerraformCliArgs[0]) {
+			commandLength = 2
+		}
+		var args []string
+		args = append(args, terragruntOptions.TerraformCliArgs[:commandLength]...)
+
+		// Options must be inserted after command but before the other args
+		// command is either 1 word or 2 words
+		args = append(args, filterTerraformExtraArgs(terragruntOptions, conf)...)
+		terragruntOptions.TerraformCliArgs = append(args, terragruntOptions.TerraformCliArgs[commandLength:]...)
```
Comment (brikis98) [inline cli/cli_app.go]: It would be more readable to use a consistent structure for building up the list of `args`:

```go
var args []string

// Options must be inserted after command but before the other args
// command is either 1 word or 2 words
args = append(args, terragruntOptions.TerraformCliArgs[:commandLength]...)
args = append(args, filterTerraformExtraArgs(terragruntOptions, conf)...)
args = append(args, terragruntOptions.TerraformCliArgs[commandLength:]...)

terragruntOptions.TerraformCliArgs = args
```

Comment (brikis98) [inline cli/cli_app.go]: Also, we need a test case for this!

File: config/config_helpers.go [hunk 6]
```diff
@@ -28,19 +52,46 @@ type EnvVar struct {
 func ResolveTerragruntConfigString(terragruntConfigString string, include *IncludeConfig, terragruntOptions *options.TerragruntOptions) (resolved string, finalErr error) {
 	// The function we pass to ReplaceAllStringFunc cannot return an error, so we have to use named error
 	// parameters to capture such errors.
-	resolved = INTERPOLATION_SYNTAX_REGEX.ReplaceAllStringFunc(terragruntConfigString, func(str string) string {
+
+	resolved = INTERPOLATION_SYNTAX_REGEX_SINGLE.ReplaceAllStringFunc(terragruntConfigString, func(str string) string {
+		// We do a special treatment for function returning an array of values because they need
+		// to be declared like that "${function()" and the result must get rid of the surrounding quote
```
Comment (brikis98) [inline config/config_helpers.go]: Might be clearer to say:

```
All interpolation functions are called using the syntax "${function_name()}". For functions that return a non-string value we have to get rid of the surrounding quotes and convert the output to HCL syntax. For example, for an array, we need to return [1, 2, 3].
```

File: config/config_helpers.go [hunk 7]
```diff
@@ -28,19 +52,46 @@ type EnvVar struct {
...
+		internalStr := str[1 : len(str)-1] // We get rid of the surrounding quote
+		out, err := resolveTerragruntInterpolation(internalStr, include, terragruntOptions)
+		if err != nil {
+			finalErr = err
+		}
+
+		switch out := out.(type) {
+		case []string:
+			return util.ListToHCLArray(out)
+		}
```
Comment (brikis98) [inline config/config_helpers.go]: There should be two `case` statements here: the one for `[]string` you already have and a new one for `string` that returns the string unchanged. There should also be a `default` that sets `finalErr` to an error message indicating that an unsupported type was returned and using `reflect.TypeOf` to display what that type was.

File: config/config_helpers.go [hunk 8]
```diff
@@ -28,19 +52,46 @@ type EnvVar struct {
...
+
+		// The function is not returning an array of string, so left the string unchanged
+		// Remaining functions will be processed by the next statement
+		return str
+	})
+
+	resolved = INTERPOLATION_SYNTAX_REGEX.ReplaceAllStringFunc(resolved, func(str string) string {
```
Comment (brikis98) [inline config/config_helpers.go]: Why are there two regex now and why should both be executed?

Comment (jocgir) [inline config/config_helpers.go]: There is one for the case where the functions is directly included within "" as `"${get_terraform_commands_that_need_vars()}"` which does not return the same result as something hypothetic like `"Hello ${get_terraform_commands_that_need_vars()} world!"`. In the first case, we return a list and in the second, we return the expansion of the list within a string.

File: README.md (hunk 4 above)
Comment (jocgir) [inline README.md]: I changed it, but in fact, this is not resulting in an error. It is just that the result would be useless. That explain why I used shall not instead of cannot.

File: config/config_helpers.go (hunk 7 above)
Comment (jocgir) [inline config/config_helpers.go]: I am not sure that this would be a plus. That would just add a restriction to future interpolation function. The goal here is to properly process the array return and let other any other values unchanged.

Then, the next statement take care of the rest.

File: README.md (hunk 4 above)
Comment (brikis98) [inline README.md]: A useless result is worse than an error, as it's an error that's going to be hard to track down :)

How about "We do not recommend using them in string composition"?

File: config/config_helpers.go (hunk 8 above)
Comment (brikis98) [inline config/config_helpers.go]: Ah, OK. TBH, this structure with two regular expressions and various shuffling of types is a bit confusing. At the very least, we need clear comments explaining how this works. I'll also think about a way to simplify this...

Comment (brikis98) [inline config/config_helpers.go]: Also, we need tests for this stuff too, as it has lots of corner cases.

File: util/collections.go [hunk 9]
```diff
@@ -49,3 +54,12 @@ func removeDuplicatesFromList(list []string, keepLast bool) []string {
 	}
 	return out
 }
+
+// Returns an HCL compliant formated list of strings
+func ListToHCLArray(list []string) string {
```
Comment (brikis98) [inline util/collections.go]: This function makes it sound like it'll return a string of the format `["foo", "bar", "baz"]`, but in reality, it returns `"foo", "bar", "baz"` (no brackets). Therefore, it should probably be called something like `ToCommaSeparatedStrings` or similar.

File: config/config_helpers.go (hunk 8 above)
Comment (brikis98) [inline config/config_helpers.go]: OK, after thinking about this for a few minutes more, it just seems that Terraform's interpolation syntax, by requiring double quotes around interpolations, is inherently messy. I don't want to waste any more time on looking for a more elegant way to parse/process it, so for now, let's add do the following:

1. Move each regex `ReplaceAllStringFunc` call into its own function. The first one can be called `processSingleInterplationInString` and the second `processMultipleInterpolationsInString`. Each one should have a comment above it with an example of what it does. E.g. the comment above `processSingleInterplationInString` would look something like this:

    ```
    Process a string that has a single interpolation in it and nothing else, such as "${function()}". These sorts of interpolations allow us to render types other than a string. For example, if "${function}" returns an array, we can drop the wrapping quotes and render [1, 2, 3].
    ```
1. Add automated tests for these new interpolation behaviors.

File: util/collections.go (hunk 9 above)
Comment (jocgir) [inline util/collections.go]: Initially, this function was returning a list ;-) but you preferred that the user add the bracket itself.

File: config/config_helpers.go [hunk 10]
```diff
@@ -49,47 +49,83 @@ type EnvVar struct {
...
+	// First, we replace all single interpolation syntax (i.e. function directly enclosed within quotes "${function()}")
+	terragruntConfigString, err := processSingleInterpolationInString(terragruntConfigString, include, terragruntOptions)
+	if err != nil {
+		return terragruntConfigString, err
+	}
+	// Then, we replace all other interpolation functions (i.e. functions not directly enclosed within quotes)
+	return processMultipleInterpolationsInString(terragruntConfigString, include, terragruntOptions)
+}
+
+// Execute a single Terragrunt helper function and return its value as a string
```
Comment (brikis98) [inline config/config_helpers.go]: s/and return its value as a string/and return the result/

File: config/config_helpers.go [hunk 11]
```diff
@@ -49,47 +49,83 @@ type EnvVar struct {
...
-	// The function we pass to ReplaceAllStringFunc cannot return an error, so we have to use named error
-	// parameters to capture such errors.
+func ResolveTerragruntConfigString(terragruntConfigString string, include *IncludeConfig, terragruntOptions *options.TerragruntOptions) (string, error) {
+	// First, we replace all single interpolation syntax (i.e. function directly enclosed within quotes "${function()}")
+	terragruntConfigString, err := processSingleInterpolationInString(terragruntConfigString, include, terragruntOptions)
+	if err != nil {
+		return terragruntConfigString, err
+	}
+	// Then, we replace all other interpolation functions (i.e. functions not directly enclosed within quotes)
+	return processMultipleInterpolationsInString(terragruntConfigString, include, terragruntOptions)
```
Comment (brikis98) [inline config/config_helpers.go]: Ah, much easier to understand now, thx!

File: config/config_helpers_test.go [hunk 12]
```diff
@@ -492,6 +493,27 @@ TERRAGRUNT_HIT","")}/bar`,
...
+			fmt.Sprintf("commands = [%s]", util.CommaSeparatedStrings(TERRAFORM_COMMANDS_NEED_VARS)),
+			nil,
+		},
+		{
+			`commands = "test-${get_terraform_commands_that_need_vars()}"`,
+			nil,
+			options.TerragruntOptions{TerragruntConfigPath: "/root/child/" + DefaultTerragruntConfigPath, NonInteractive: true},
+			fmt.Sprintf(`commands = "test-%v"`, TERRAFORM_COMMANDS_NEED_VARS),
+			nil,
+		},
```
Comment (brikis98) [inline config/config_helpers_test.go]: Nice tests! Could you add one more that calls two interpolation functions in one string?

**Comment by brikis98** [comment]
OK, we are almost there. Two final comments and this is good to merge. Thank you for this important fix!

**Comment by jocgir** [comment]
You are right, more tests we have, more problems we can identify. Read my latest commit to check what problem I faced :-)

**Comment by brikis98** [comment]
Hahah, awesome. Merging now. Thanks!

**Comment by brikis98** [comment]
New build is here: https://github.com/gruntwork-io/terragrunt/releases/tag/v0.12.17
